### ISSUE TITLE
Leak in future::then

### ISSUE THREAD

**Issue by lsr0** [RFerraro/eventual#11]
Debugging a leak issue led to the discovery that callbacks passed to `eventual::shared_future::then()` were never cleaned up. Further investigation suggests some parts of the State chain are retained whole. Below is a demonstrative test case:

```cpp
#include <eventual/eventual.h>
#include <cstdio>
#include <memory>

static int canary_count = 0;
struct canary
{
    canary()  { canary_count += 1; printf("+canary (%d)\n", canary_count); }
    ~canary() { canary_count -= 1; printf("-canary (%d)\n", canary_count); }

    char block[1*1024*1024*1024];
};

static void promise_scope()
{
    eventual::promise<void> promise;
    auto fut = promise.get_future();

    auto shared_canary = std::make_shared<canary>();
    fut.then([shared_canary] (auto&) {});

    promise.set_value();
}

int main()
{
    promise_scope();
    return 0;
}
```

The output of the above is
`+canary (1)`
and valgrind indicates:
`in use at exit: 1,073,742,464 bytes in 6 blocks`

One fix seems to be explicitly clearing the callback list after invocation, but I do not know whether I am just working around a deeper problem, so have not created a pull request. The change is visible in lsr0/eventual@fe0139927462cf33a7cf0cb30aba4daf053fdc29 for reference.

**Comment by RFerraro** [comment]
If your change works as described, it seems to imply that either std::list is not releasing the callback nodes as expected, or detail::State<T> is outliving the scope. However, I won't have time to take a deeper look into this until early next week... 

In the meantime, can you give me details about the toolchain you're seeing this problem in? 
Specifically: operating system, compiler, standard library, and their versions.

**Comment by lsr0** [comment]
Sure. All on linux, options tried:
* gcc 7.3.0 + libstdc++ (from gcc 7.3.0)
* gcc 5.5.0 + libstdc++ (from gcc 5.5.0)
* clang 7.0.0 (r326682) + libstdc++ (from gcc 7.3.0)
* clang 7.0.0 (r326682) + libc++ 5.0.0

**Comment by RFerraro** [comment]
I managed to look into the issue a little bit more, and found the root cause is a circular reference between the detail::State and it's continuations (which hold a shared_pointer of detail::State). So the reason why your change [lsr0/eventual@fe01399](https://github.com/lsr0/eventual/commit/fe0139927462cf33a7cf0cb30aba4daf053fdc29) works is it's breaking the circular reference.

It may also be better to clear the continuations as they are invoked, something like:
```cpp
auto current = _callbacks.begin();
while (current != _callbacks.end())
{
   (*current)(); // todo: exceptions?
   current++;
   _callbacks.pop_front(); // break circular references.
}
```

I'll have additional time to look into this bug early next week, and since you've provided a concise test case I'd like to include a variant of it into the unit test suite.

**Comment by RFerraro** [comment]
Fixed via #15
